Newsletters: add 259 (2023-07-12) (#1183)
bitcoinops/bitcoinops.github.io: _includes/specials/policy/en/09-proposals.md

[Last week’s post][policy08] described anchor outputs and CPFP carve
out, ensuring either channel party can fee-bump their shared
commitment transactions without requiring collaboration. This approach
still contains a few drawbacks: channel funds are tied up to create
anchor outputs, commitment transaction feerates typically overpay
to ensure they meet mempool minimum feerates, and CPFP carve out only

> allows one extra descendant. Anchor outputs do not ensure the same

Reviewer: ```suggestion
allows one extra descendant. Anchor outputs cannot ensure the same
```

ability to fee-bump for transactions shared between more than two
parties, such as CoinJoins or multi-party contracting protocols. This
post explores current efforts to address these and other limitations.

[Package relay][topic package relay] includes p2p protocol and policy
changes to enable the transport and validation of groups of

> transactions. It would allow a commitment transaction to be fee-bumped
> by a child even if a mempool’s minimum feerate exceeds that of the
> commitment transaction. Additionally, _Package RBF_ would allow

Reviewer: ```suggestion
transactions. It would allow a commitment transaction to be fee-bumped
by a child even if the commitment transaction does not meet a mempool’s minimum feerate. 
Additionally, _Package RBF_ would allow
```

the fee-bumping child to pay for replacing transactions its parent

> conflicts with. Package relay is not designed with any particular use
> case in mind; it removes a general limitation at the base protocol
> layer. However, due to its utility in fee-bumping of shared

Reviewer: Maybe:

```suggestion
conflicts with. Package relay is not designed with just one particular use
case in mind; it removes a general limitation at the base protocol
layer. However, due to its utility in fee-bumping of shared
```

by all to-be-replaced transactions. This rule helps prevent DoS
through repeated replacements, but allows a malicious user to increase
the cost to replace their transaction by attaching a child that is
high fee but low feerate. This hinders the transaction from being
mined by unfairly preventing its replacement by a high-feerate
package, and is often referred to as “Rule 3 pinning.”

Developers have also proposed entirely different ways of adding fees
to presigned transactions. For example, signing inputs of the
transaction using `SIGHASH_ANYONECANPAY | SIGHASH_ALL` could allow the
transaction broadcaster to provide fees by appending additional inputs
to the transaction without changing the outputs. However, as RBF does
not have any rule requiring the replacement transaction to have a
higher “mining score” (i.e. would be selected for a block faster), an
attacker could pin these types of transactions by creating
replacements encumbered by low-feerate ancestors. What complicates
the accurate assessment of the mining score of transactions and
transaction packages is that the existing ancestor and descendant
limits are insufficient to bound the computational complexity of this

> calculation. Any transactions it is connected to can influence the
> order in which transactions get picked into a block. A fully-connected

Reviewer: Maybe:

```suggestion
calculation. Any connected transactions can influence the
order in which transactions get picked into a block. A fully-connected
```

component, called a _cluster_, can be of any size given current
ancestor and descendant limits.

A long term solution to address some mempool deficiencies and RBF
pinning attacks is to [restructure the mempool data structure to track
clusters][mempool clustering] instead of just ancestor and descendant
sets. These clusters would be limited in size. A cluster limit would
restrict the way users can spend unconfirmed UTXOs, but make it
feasible to quickly linearize the entire mempool using the ancestor
score-based mining algorithm, build block templates extremely quickly,
and add a requirement that replacement transactions have a
higher mining score than the to-be-replaced transaction(s).

Even so, it’s possible that no single set of policies can meet the
wide range of needs and expectations for transaction relay. For

> example, while batched payments may require a number of unconfirmed

Author: I don't understand "while batched payments may require a number of unconfirmed descendants, this leaves room for pinning"

Is it supposed to be s/while/when/?   Or is "batched payments" referring to something besides https://bitcoinops.org/en/topics/payment-batching/ ?

descendants, this leaves room for pinning package RBF of a shared
transaction through absolute fees. A proposal for [v3 transaction
relay policy][topic v3 transaction relay] was developed to allow
contracting protocols to opt in to a more restrictive set of package
limits. V3 transactions would only permit packages of size two (one
parent and one child) and limit the weight of the child. These limits

> would mitigate RBF pinning through absolute fees, and offer some of

Reviewer: ```suggestion
would delimit RBF pinning through absolute fees, and offer some of
```

Author: I don't see a problem with "mitigate" and I think it's better than "these limits would delimit [...]".

the benefits of cluster mempool without requiring a mempool
restructure.

[Ephemeral Anchors][topic ephemeral anchors] builds upon the
properties of v3 transactions and package relay to improve anchor

> outputs further. It removes the dust limit for an anchor output
> belonging to a zero-fee v3 transaction, provided the anchor output is

Reviewer: ```suggestion
outputs further. It exempts anchor outputs
belonging to a zero-fee v3 transaction from the dust limit, provided the anchor output is
```

> spent by a fee-bumping child.  Since the zero-fee transaction must be
> fee-bumped by exactly one child (otherwise a miner would not be
> incentivized to include it in a block), this anchor output is
> “ephemeral” and would not become a part of the UTXO set. The ephemeral

Reviewer: ```suggestion
spent by a fee-bumping child.  Since there would be no incentive to include a zero-fee transaction in a block, it must be
fee-bumped by a child of which there can be only one due to the v3 rules. This prevents the “ephemeral” anchor output
from becoming a part of the UTXO set. The ephemeral
```

Do I recall correctly that the ephemeral anchors proposal would also prescribe that all other outputs on the commitment transactions would be CSV-locked to be unspendable before confirmation?

Author: @murchandamus 

I think it's the opposite: "Since this anchor MUST be spent, any spending of other outputs in the
same parent transaction MUST directly double-spend prior spends of the
ephemeral anchor. This causes the 1 block CSV timelock on outputs to be
removed in these situations."  -- https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-October/021036.html

Greg's phrasing above was bit unclear to me on first reading, so in my own words: if Alice and Bob have a v3 commitment transaction with a 0-value output (ephemeral anchor), the commitment transaction will only be accepted by relay nodes if it's bundled with a spend of that output.  There's no other way under the v3 rules and the ephemeral anchor rules for the transaction to be accepted for relay.  The only alternative to Alice and Bob would be to broadcast a mutual close that doesn't include a zero-value output.

I think you might be thinking of the existing anchor outputs where 1 OP_CSV is used.

Reviewer: Thanks for the correction!
